### ISSUE TITLE
Keras Mask Issue When Loading Model With Custom Layer

### ISSUE THREAD

**Issue by jonas-eschmann** [keras-team/keras#15510]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab / Ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.6.0-0-g919f693420e 2.6.0 (Colab) / 2.5.0 (Ubuntu 18.04)
- Python version: major=3, minor=7, micro=12, releaselevel='final', serial=0
- Bazel version (if compiling from source):
- GPU model and memory: K80 12GB
- Exact command to reproduce: Colab Run All

**Describe the problem**.

Custom layers can opt in to support the `_keras_mask`, but when a custom layer is part of a loaded model and the layer source code is not explicitly provided, tensorflow/keras will silently fails and creates a model that does produce incorrect results.
The missing reproduction of the "supports_masking" property leads to extremely nasty errors that are hard to track down for the target audience.

**Describe the current behavior**.
Loaded model output differs from the saved one (without any warning).

**Describe the expected behavior**.
Giving a warning, when a model with custom layers that "support_masking" is saved or incorporating the masking into the saved tensorflow graph.

- Do you want to contribute a PR? (yes/no): no

**Standalone code to reproduce the issue**.
Colab: https://colab.research.google.com/drive/1GDX7H08cR_dbzJCzlA2faUU3BA2yDAWf?usp=sharing
```
import tensorflow as tf
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Layer, LSTM, Masking
from tensorflow.keras.models import load_model
import numpy as np
import os, sys
print(tf.version.GIT_VERSION, tf.version.VERSION)
print(sys.version_info)
!nvidia-smi

class CustomLayer(Layer):
    def __init__(self, **kwargs):
        super(CustomLayer, self).__init__(**kwargs)
        self.supports_masking = True

    def call(self, inputs):
        return inputs

    def get_config(self):
        config = super(CustomLayer, self).get_config()
        return config

model_save_path = os.path.join("tmp", "keras_model")
N = 10000
seq_len = 50
X = np.array([(b + 0.1 * np.random.randn()) * (np.arange(seq_len) < l) for l, b in zip(np.random.randint(1, seq_len, N), np.random.randn(N))])
X = X[:, :, np.newaxis]
Y = X.mean(axis=-2)

X_train = X[:-1000, :, :]
Y_train = Y[:-1000, :]
X_test = X[-1000:, :, :]
Y_test = Y[-1000:, :]

model = Sequential([Masking(), CustomLayer(), LSTM(1)])

model.compile(loss="mse")
model.fit(X_train, Y_train, batch_size=32, epochs=30)
model.save(model_save_path)

model_loaded = load_model(model_save_path)
custom_objects = {"CustomLayer": CustomLayer}
with tf.keras.utils.custom_object_scope(custom_objects):
    model_loaded_properly = load_model(model_save_path)


for (res, res_loaded, res_loaded_properly, gt) in zip(model(X_test[:10]), model_loaded(X_test[:10]), model_loaded_properly(X_test[:10]), X_test[:10].mean(axis=(-2))):
  print(f"res: {res} res_loaded: {res_loaded} res_loaded_properly: {res_loaded_properly} ground truth: {gt}")
```

**Source code / logs**.
```
res: [-0.20905927] res_loaded: [-0.26638848] res_loaded_properly: [-0.20905927] ground truth: [-0.1800429]
res: [-0.78159744] res_loaded: [-0.797364] res_loaded_properly: [-0.78159744] ground truth: [-0.82082076]
res: [-0.8153708] res_loaded: [-0.818481] res_loaded_properly: [-0.8153708] ground truth: [-0.8807303]
res: [0.7262878] res_loaded: [0.734436] res_loaded_properly: [0.7262878] ground truth: [0.70544283]
res: [-0.05942882] res_loaded: [-0.06771344] res_loaded_properly: [-0.05942882] ground truth: [-0.05173485]
res: [0.8717314] res_loaded: [0.867892] res_loaded_properly: [0.8717314] ground truth: [1.07998851]
res: [0.5259318] res_loaded: [0.5569656] res_loaded_properly: [0.5259318] ground truth: [0.44405287]
res: [-0.01080081] res_loaded: [-0.00134738] res_loaded_properly: [-0.01080081] ground truth: [-0.01114861]
res: [-0.12931338] res_loaded: [-0.1299946] res_loaded_properly: [-0.12931338] ground truth: [-0.10491925]
res: [-0.45746583] res_loaded: [-0.53842115] res_loaded_properly: [-0.45746583] ground truth: [-0.40699462]
```

**Comment by jonas-eschmann** [comment]
@jvishnuvardhan thanks for your answer! But I know this (I found out the hard way). The problem is that this is not stated in the docs and just works, but produces wrong results. This is extremely confusing to the user, because even the most complex tf ops that might happen within the layer are saved as part of the graph, but then a kind of mundane thing like the _keras_mask does not work. I think this has to have led to many bugs, frustration and hours/days wasted for people, but I'm also not sure how to fix it besides printing a warning, when a custom layer supporting masking is saved
